### PR TITLE
WIP: Add a lint that detects non-portable usize literals

### PR DESCRIPTION
**Pull request by alexanderkjall**
First lint I write, so I thought I should start with a minimal version and get feedback on it before spending the time fleshing it out.

I'm doing some packaging of rust crates for Debian in my spare time, and that results in running the builds and unit-tests for crates on more architectures than the author typically have access to, and there are a number of common mistakes that makes code nonportable.

Not thinking about `usize` and `isize` sizes are maybe the most common error.

One example of the issues that I would like to prevent is here: https://gitlab.com/sequoia-pgp/sequoia-chameleon-gnupg/-/issues/41

Since this codebase is very new to me, I would love feedback on everything.

I also realize that there are a large number of cases that isn't caught by this lint, but as mentioned before, I would rather get an early round of feedback.

---

changelog: [`usize_unportable_32_bit_literal`]: Add a lint that detects usize assignments that are not portable to 32bit architectures

### CODE REVIEW

**Comment by PatchMixolydic** [comment]
On that note, there should probably be a separate pedantic/restriction lint for code that isn't portable to targets with [16-bit `{u,i}size`](https://doc.rust-lang.org/reference/types/numeric.html#machine-dependent-integer-types).

**Comment by alexanderkjall** [comment]
yes, I agree, I think that might be a quite different usecase as 16 bit imposes a lot tighter restrictions, so it would make sense to have them as sepparate lints.

File: clippy_lints/src/usize_unportable_32_bit_literal.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    while Some(enclosing_node) != prev_enclosing_node {
+        if has_pointer_width_64_attr(map.attrs(enclosing_node)) {
+            return true;
+        }
+        prev_enclosing_node = Some(enclosing_node);
+        enclosing_node = map.get_parent_item(enclosing_node).into();
+    }
+
+    false
+}
```
Comment (Jarcho) [inline clippy_lints/src/usize_unportable_32_bit_literal.rs:129]: This should use `map.parent_iter().any(...)`. Statements can have attributes, so you can't just use the parent item. Expression cannot have attributes, so you can skip the initial node (the function name also implies you do this).

File: clippy_lints/src/usize_unportable_32_bit_literal.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    /// ```no_run
+    /// let foo : usize = ;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let foo : u64 = 1<<32;
+    /// ```
+    #[clippy::version = "1.78.0"]
+    pub USIZE_UNPORTABLE_32_BIT_LITERAL,
+    correctness,
```
Comment (Jarcho) [inline clippy_lints/src/usize_unportable_32_bit_literal.rs:35]: This should be suspicious at most. We really need a portability category.

**Review by Jarcho — CHANGES_REQUESTED** [review]

**Comment by J-ZhengLi** [comment]
Hey @alexanderkjall, this is a ping from triage.

This PR is labeled WIP but there hasn't been any activity for some time, therefore I'm just here to make sure that you're still working on it and not stuck. 😃

**Comment by alexanderkjall** [comment]
@J-ZhengLi Thanks for the ping, I still intend to see this through, but have gotten very busy with some other things in my life the last two months and that will continue for another one or so. No hard feelings if you want to close this or someone else want to work on the idea :)

**Comment by J-ZhengLi** [comment]
> ...I still intend to see this through, but have gotten very busy with some other things in my life the last two months and that will continue for another one or so. No hard feelings if you want to close this or someone else want to work on the idea :)

That's very understandable~ I'll go ahead close this for now, but you can always re-open it when you got time.

You may open a new issue on this idea and assign yourself with `@rustbot claim` if you want to "lock it" 😄. If you need more help, you are welcome to ask people in the [clippy stream on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy).

@rustbot label +S-inactive-closed

**Pull request closed without merging**